### ISSUE TITLE
[전 챕터][EPUB][Major] notebook_to_tex 가 코드 안의 그리스문자·화살표까지 $...$ 로 치환 — 19개 챕터 62곳, #137·#145 의 공통 원인

### ISSUE THREAD

**Issue by jwrim9305** [yoon-gu/neuqes-101#169]
## 대상

- **도구**: `book/tools/notebook_to_tex.py:2010` `normalize_markdown_math_symbols()`
- **영향**: tex 를 재생성하는 모든 챕터
- **기준 커밋**: `4863c9e` (master)

## 실제 결과

master 의 tex 원고 **19개 챕터에 62곳** 이 이 방식으로 깨져 있습니다. 도구가 그렇게 만들어 냅니다.

```bash
python3 - <<'EOF'
import re, glob, os
sym = 'lambda|to|alpha|beta|theta|Delta|approx|le|ge|times|ne|leftarrow|downarrow|leftrightarrow'
pat = re.compile(r'\\?\$\\textbackslash(?:\{\}|\s+)(' + sym + r')\\?\$')
tot = chs = lam = 0
for f in sorted(glob.glob('book/chapters/*.tex')):
    m = pat.findall(open(f).read())
    if m:
        chs += 1; tot += len(m); lam += m.count('lambda')
        print('%3d (λ %d)  %s' % (len(m), m.count('lambda'), os.path.basename(f)))
print('영향 챕터', chs, '| 총', tot, '| λ', lam)
EOF
# → 영향 챕터 19 | 총 62 | λ 19
```

함수 단위로 재현됩니다.

```python
>>> normalize_markdown_math_symbols("`loss = main + λ * aux`")
'`loss = main + $\\lambda$ * aux`'          # ← 코드 스팬 안인데 치환됨

>>> normalize_markdown_math_symbols("산문에서 λ 는 정상입니다")
'산문에서 $\\lambda$ 는 정상입니다'            # ← 여기서는 올바른 동작

>>> normalize_markdown_math_symbols("```\nlr → 0  # α 조정\n```")
'```\nlr $\\to$ 0  # $\\alpha$ 조정\n```'    # ← 펜스 블록 안까지, λ 이외 기호도
```

**산문에서는 올바르고 코드 안에서만 틀립니다.** 코드에서는 `$...$` 가 수식이 아니라 그냥 문자이기 때문입니다.

## 원인

```python
# book/tools/notebook_to_tex.py:2011
def normalize_markdown_math_symbols(text: str) -> str:
    return (
        text.replace("λ", r"$\lambda$")
        .replace("β", r"$\beta$")
        .replace("α", r"$\alpha$")
        ...
        .replace("→", r"$\to$")
        .replace("≈", r"$\approx$")
        ...
    )
```

마크다운 **텍스트 전체에 무조건** `str.replace` 를 겁니다. 산문에서는 맞지만, 코드 표기 안에서는 LaTeX 이 `$...$` 를 수식으로 읽지 않아 **글자 그대로** 노출됩니다.

| 기호가 있는 자리 | LaTeX 이 `$\lambda$` 를 만나면 | 결과 |
|---|---|---|
| 산문 | 수식으로 해석 | **λ** ✅ |
| `\texttt{}` · `lstlisting` (코드) | 해석하지 않고 그대로 출력 | **`$\lambda$`** ❌ |

### 깨지는 과정 — 실제 사례 셋

**① Ch 12 파라미터 표**

```
노트북   | pre_classifier (`Linear(768→768)`) | 590,592 | 590,592 |
tex      pre\_classifier (\texttt{Linear(768\$\textbackslash{}to\$768)}) & 590,592 ...
책       pre_classifier (Linear(768$\to$768))          ← 화살표가 글자 6개로
```

**② Ch 14 절 제목**

```
노트북   ## 손실 노트 — Combined loss `L = L_main + λ · L_aux`
tex      \section{... \texttt{L\ =\ L\_main\ +\ \$\textbackslash{}lambda\$\ ·\ L\_aux}}
책       손실 노트 — Combined loss L = L_main + $\lambda$ · L_aux
```

**③ Ch 26** (λ 가 없는 챕터)

```
노트북   `ln(2048) ≈ 7.62` 와 같은 직관
책       ln(2048) $\approx$ 7.62 와 같은 직관
```

### 영향 범위 — 헷갈리기 쉬운 구분

**어떤 산출물이 영향을 받나**

| 산출물 | 경로 | 영향 |
|---|---|---|
| **책 (EPUB·PDF)** | `notebook_to_tex.py` → tex | ❌ **여기만 깨집니다** |
| WikiDocs (`pages/*.md`) | `build_wikidocs.py` | ✅ 무관 — 이 함수를 거치지 않음 |
| 노트북 | 원본 | ✅ 무관 |

**노트북 어디가 영향을 받나**

| 자리 | 영향 | 예 |
|---|---|---|
| 마크다운 셀의 **인라인 코드 스팬** | ❌ | `` `L = L_main + λ·L_aux` `` |
| 마크다운 셀의 **표 안 코드** | ❌ | `` | `Linear(768→768)` | 590,592 | `` |
| 마크다운 셀의 **절 제목 안 코드** | ❌ | ``## 손실 노트 — `L = ... + λ · L_aux` `` |
| 마크다운 셀의 **펜스 블록** | ❌ | ` ```python ... ``` ` |
| 마크다운 셀의 **산문** | ✅ | 「λ 를 키우면 …」 — 정상 동작 |
| **실행되는 코드 셀** | ✅ | 다른 경로로 처리됨 |

즉 **「마크다운 셀에 백틱으로 적은 코드」** 가 대상입니다. 주석만이 아니라 표 셀·절 제목까지 포함되고, 실제로 Ch 14 에서 깨진 6곳에 절 제목과 표 셀이 들어 있습니다.

**λ 만의 문제도 아닙니다.** 같은 함수가 `→ ← ≈ ≤ ≥ × ≠ α β θ Δ ↓ ↔` 를 동일하게 치환합니다. 전체 62곳 중 λ 는 19곳뿐이고, **λ 가 아예 없는 챕터 12개도 깨져 있습니다.**

## 왜 도구를 고쳐야 하는가

이 결함은 지금까지 **증상으로만 두 번 보고됐습니다.** 세 사람이 각자 다른 조각을 만났습니다.

| 이슈 | 보고자 | 본 것 | 상태 |
|---|---|---|---|
| [#137](https://github.com/yoon-gu/neuqes-101/issues/137) | @sy2399 | Ch 14 의 **λ** 6곳 | 종결 (tex 손수정) |
| [#145](https://github.com/yoon-gu/neuqes-101/issues/145) | @yum-yeom | Ch 1·18·21 의 **×** 5곳 | 열림 |
| **이 이슈** | @jwrim9305 | **함수 1개 → 19챕터 62곳** | — |

앞의 둘은 같은 함수 한 줄에서 나온 **같은 결함의 다른 기호**입니다. 아무도 도구까지 내려가지 않아 매번 새로 발견되고, 매번 tex 를 손으로 고치고 있습니다.

> `#137`·`#145` 를 대체하자는 뜻이 아닙니다. 두 이슈는 각자 지목한 자리의 처리로 두고, 이 이슈는 **재발을 멈추는 것**만 다룹니다.

@sy2399 님이 손수정의 한계를 `#137` 종결 코멘트에 정확히 적어 두셨습니다.

> ⚠️ 근본 원인인 `notebook_to_tex.py:2012` 는 그대로입니다. **tex 를 재생성하면 전부 재발합니다.** tex 손수정이 맞습니다.

문제는 **tex 재생성이 선택 사항이 아니라는 점**입니다. tex 는 `--use-executed` 로 만들어지므로, 검수에서 **재실행한 챕터는 반드시 tex 를 다시 생성해야** 합니다. 즉 「검수에서 재실행한다」와 「tex 손수정으로 λ 를 유지한다」가 구조적으로 양립하지 않습니다. 재실행하는 챕터마다 손수정을 다시 입혀야 하고, 한 번이라도 빠뜨리면 조용히 깨집니다.

`#137` 코멘트에는 **λ 만** 센 분포가 정리돼 있는데, 기호 전체로 넓히면 규모가 크게 달라집니다.

| | `#137` 코멘트 (λ 만) | 실측 (기호 전체) |
|---|---|---|
| 영향 챕터 | 7 | **19** |
| 깨진 자리 | 21 | **62** |

챕터별 분포입니다.

```
  7 (λ 0)  ch01_tfidf              ← λ 가 없는데도 깨짐 (× ≈)
  2 (λ 0)  ch07_bert_pipeline
  1 (λ 0)  ch08_tokenizer_datasets
  2 (λ 0)  ch11_bert_binary_softmax
  4 (λ 0)  ch12_bert_multiclass    ← Linear(768$\to$768)
  3 (λ 2)  ch13_bert_multilabel
  6 (λ 6)  ch14_auxiliary_loss
  2 (λ 2)  ch17_ko_multilabel
  7 (λ 6)  ch18_ko_auxiliary
  1 (λ 1)  ch19_tokenizer_training
  4 (λ 1)  ch20_en_bert_pretrain
  2 (λ 1)  ch21_en_bert_classify
  1 (λ 0)  ch22_ko_bert_pretrain
  2 (λ 0)  ch24_gpt_tinystories
  4 (λ 0)  ch25_gpt2_continual_pretrain
  6 (λ 0)  ch26_ko_tiny_gpt        ← ln(2048)$\approx$7.62
  4 (λ 0)  ch27_ko_gpt2_continual_pretrain
  2 (λ 0)  ch30_dpo
  2 (λ 0)  ch32_diffusion_intro
```

**λ 는 62 중 19 뿐입니다.** 나머지 43 은 `→ × ≈` 등이고, **λ 가 아예 없는 챕터도 12개** 깨져 있습니다. 지면에 이렇게 실립니다.

```latex
ch12:  pre_classifier (Linear(768$\textbackslash{}to$768))
ch26:  ln(2048) $\textbackslash{}approx$ 7.62
ch01:  5000 $\textbackslash{}times$ 10000 $\textbackslash{}times$ 8
```

## 기대 결과

**코드 구간을 보호한 뒤 치환합니다.** 펜스·인라인 코드 스팬을 먼저 떼어 두고, 바깥 텍스트에만 적용한 다음 되돌립니다.

```python
import re

_CODE_SEGMENT = re.compile(r'(```.*?```|~~~.*?~~~|`[^`\n]+`)', re.DOTALL)

_MATH_SYMBOLS = {
    "λ": r"$\lambda$", "β": r"$\beta$", "α": r"$\alpha$", "θ": r"$\theta$",
    "Δ": r"$\Delta$", "≈": r"$\approx$", "≤": r"$\le$", "≥": r"$\ge$",
    "×": r"$\times$", "→": r"$\to$", "←": r"$\leftarrow$",
    "↓": r"$\downarrow$", "↔": r"$\leftrightarrow$", "≠": r"$\ne$",
}


def normalize_markdown_math_symbols(text: str) -> str:
    """코드 스팬·펜스 밖의 수학 기호만 $...$ 로 바꾼다."""
    def convert(segment: str) -> str:
        for src, dst in _MATH_SYMBOLS.items():
            segment = segment.replace(src, dst)
        return segment.replace("−", "-")

    # 홀수 인덱스가 코드 구간 — 그대로 두고 짝수 인덱스만 변환
    parts = _CODE_SEGMENT.split(text)
    return "".join(
        part if i % 2 else convert(part) for i, part in enumerate(parts)
    )
```

`−`(U+2212 → 하이픈)는 코드 안에서도 안전하므로 양쪽 모두 적용합니다.

### 코드 구간에 남는 리터럴 글리프 처리

치환을 막으면 코드 안에 **리터럴 λ** 가 남습니다. 폰트 지원이 갈립니다 — `#137` 작업에서 확인된 내용입니다.

| 폰트 | 용도 | λ (U+03BB) |
|---|---|---|
| NanumGothicCoding | 모노 (`\texttt`, `lstlisting`) | **있음** |
| NanumMyeongjo | 본문·표 캡션 | **없음** → 두부 |

따라서 **표 캡션(`\begin{booktable}{...}`)만 `\ensuremath{\lambda}`** 로 따로 처리하면 됩니다. 나머지 코드 구간은 리터럴 λ 로 두어도 안전합니다.

## 재현 절차

```bash
git checkout 4863c9e   # master
python3 - <<'EOF'
import re, pathlib
src = pathlib.Path('book/tools/notebook_to_tex.py').read_text()
m = re.search(r'def normalize_markdown_math_symbols.*?\n(?=\n\ndef )', src, re.DOTALL)
ns = {}; exec(m.group(0), ns)
f = ns['normalize_markdown_math_symbols']
print(f("`loss = main + λ * aux`"))
print(f("```\nlr → 0  # α 조정\n```"))
EOF
```

```
`loss = main + $\lambda$ * aux`
```
lr $\to$ 0  # $\alpha$ 조정
```
```

## 전파 범위

- `book/tools/notebook_to_tex.py` — 함수 하나 교체
- 수정 후 **영향 19개 챕터 tex 재생성**으로 손수정분을 도구가 대체
- 각 챕터 EPUB·PDF 재빌드
- **재실행 불필요** — tex 변환 단계만 바뀝니다

## 심각도

**Major 로 제안합니다.**

- **이미 종결된 결함([#137](https://github.com/yoon-gu/neuqes-101/issues/137))이 정상 절차만 밟아도 되살아납니다.** 재발 조건이 「검수에서 재실행한 챕터」라 앞으로 계속 발생합니다.
- **19개 챕터 tex 62곳** 이 같은 손수정에 의존하고 있어, 누구든 재생성하는 순간 조용히 깨집니다.
- 지면(EPUB·PDF)에 그대로 실리는 렌더링 결함입니다.
- 수정 범위는 **함수 하나**로 작고, 고치면 19개 챕터의 손수정 부채가 한 번에 정리됩니다.

## 담당

- 수정 담당자: 미정 (도구 소유자)
- 재검수자: 등록자 (@jwrim9305), @sy2399, @yum-yeom
- 연결: 증상 [#137](https://github.com/yoon-gu/neuqes-101/issues/137) (종결) · [#145](https://github.com/yoon-gu/neuqes-101/issues/145) (열림) / 챕터 검수 이슈 [#38](https://github.com/yoon-gu/neuqes-101/issues/38) · [#37](https://github.com/yoon-gu/neuqes-101/issues/37)

> 이 이슈는 특정 챕터가 아니라 **도구 결함**이라 별건으로 올립니다. `#137` 을 다시 여는 대신 근본 원인을 따로 추적하는 편이 낫다고 판단했습니다 — 이견 있으면 맞추겠습니다.

**Comment by jwrim9305** [comment]
@fluentmin 님, 이 건을 맡아 주실 수 있을지 여쭙습니다.

**도구 소스(`book/tools/notebook_to_tex.py`) 수정이 필요한 건**이라 챕터 검수자가 임의로 손대기 어렵습니다. 19개 챕터 tex 가 함께 재생성되어야 해서 파급이 전 챕터에 걸치고, 이미 다른 검수자들의 진행에도 물려 있습니다.

### 지금 막혀 있는 것

| 이슈 | 보고자 | 상태 |
|---|---|---|
| #137 | @sy2399 | tex 손수정으로 종결 — **재생성하면 되돌아감** |
| #145 | @yum-yeom | 열림 — Ch 1·18·21 5곳 손수정 대기 |
| **Ch 14 (#38)** | @jwrim9305 | **재실행·tex 재생성을 이 건 대기로 보류 중** |

Ch 14 는 브랜치 머지까지 마치고 재생성 단계에서 멈춰 있습니다. 이 상태로 tex 를 재생성하면 #137 이 종결된 채로 같은 결함이 산출물에 되살아나서, 순서상 이 건이 먼저 들어가야 합니다.

### 작업 규모

**함수 하나 교체입니다.** 본문 「기대 결과」에 수정안을 넣어 두었습니다.

- `normalize_markdown_math_symbols()` — 코드 스팬·펜스를 정규식으로 분리한 뒤 바깥 텍스트에만 치환
- 표 캡션의 λ 만 `\ensuremath{\lambda}` 로 별도 처리 (본문 폰트에 글리프 없음 — #137 에서 확인된 사항)

이후 **영향 19개 챕터 tex 재생성 + EPUB 재빌드**가 따릅니다. **재실행은 불필요합니다** — tex 변환 단계만 바뀌므로 노트북 출력에 영향이 없습니다.

### 수정 후 검수

진행하시게 되면 제가 아래를 확인하겠습니다.

- [ ] 19개 챕터 tex 에서 리터럴 기호 잔여 0 (본문 재현 스크립트로 집계)
- [ ] 산문의 λ·→·≈ 등은 정상 렌더링 유지 (과잉 수정 여부)
- [ ] #137·#145 가 지목한 자리(Ch 14 6곳 · Ch 1·18·21 5곳)가 손수정 없이 해소되는지
- [ ] 코드 스팬 안 리터럴 글리프가 모노 폰트에서 정상 표시되는지

적임자가 따로 계시면 알려주세요. 맡을 분이 없으면 제가 수정해 PR 을 올리는 것도 가능한데, 도구는 전 챕터 공용이라 소유자 확인을 받는 편이 안전할 것 같아 먼저 여쭙니다.

**Comment by jwrim9305** [comment]
**#172 (@yum-yeom · `fix/epub-inline-code-math-2026-09`) 가 이 건을 해결합니다.** 중복이므로 닫습니다.

등록 시점에 해당 브랜치·PR 을 확인하지 못했습니다. 같은 근본 원인을 두 사람이 각자 추적한 셈입니다.

### 확인 결과

수정 방식이 이 이슈의 「기대 결과」와 사실상 동일합니다 — 코드 구간을 정규식으로 분리한 뒤 바깥 텍스트에만 치환.

```python
_CODE_REGION_RE = re.compile(r"(```[\s\S]*?```|~~~[\s\S]*?~~~|`[^`\n]+`)")

def normalize_markdown_math_symbols(text: str) -> str:
    parts = _CODE_REGION_RE.split(text)
    return "".join(seg if i % 2 else _normalize_math_segment(seg)
                   for i, seg in enumerate(parts))
```

**본문에서 집계한 62곳이 전부 해소됐습니다.**

| | 리터럴 기호 |
|---|---|
| master (`4863c9e`) | 19개 챕터 **62곳** |
| `fix/epub-inline-code-math-2026-09` | **0곳** |

본문의 집계 스크립트를 그대로 돌린 결과입니다. #172 쪽은 102곳으로 세셨는데, 제 패턴이 좁았던 것으로 보입니다.

### 남긴 것

표 캡션의 λ 처리에 대해 #172 에 리뷰 코멘트를 남겼습니다. 그 건은 거기서 이어가겠습니다.
